quota: new package for managing resource groups

anonymouse64 opened the pull request:
Currently, resource groups can only manage memory. The data structures here are
intended to be tracked by managers in the overlord, such that the source of
truth for what resource groups are in effect is something backed by snapd's
state.json.

Quota groups are now only applicable to entire snaps, not specific snap services - that is
to say that a single snap is in at most one quota group, and all services in that snap are
in the same quota group.

anonymouse64 commented:
You can see how this package gets used from the wrappers package at least here: https://github.com/snapcore/snapd/pull/10153

pedronis reviewed (commented):
> Additionally, it is expected that quota groups are attached to specific snaps
or apps, rather than the inverse since the quota group implementation does not
require knowing which snaps or services are present in the group, but we do
need to know when generating snap service units what quota group the service is
in.

this in principle is true but I'm not sure we will get away  not having to track groups also on their own, because we might have to allow for convenience/completeness to let a user create a group before adding any snaps to it.

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+		return fmt.Sprintf("%s.slice", grp.Name)
+	}
+
+	// otherwise we need to track back to get all of the parent elements
```
ah, this is how systemd does slice parenting/inclusion. so ok. On our side we still want the other property though.

anonymouse64 replied:
yes, names will be unique for each quota resource group, hence the check in the validate function that a sub-group cannot have the same name of the parent

but as you discovered, in order to properly implement nesting groups with proper accounting of sub-groups in the parent group, systemd needs them to be nested like this

pedronis replied:
I'm not sure if " names will be unique for each quota resource group" means globally unique, but the latter is what we want.

anonymouse64 replied:
yes sorry I meant that all quota resource groups on a given machine will be globally unique

mvo5 commented on quota/quota_test.go:
```diff
@@ -0,0 +1,215 @@
+	c.Assert(sub2.SliceFileName(), Equals, "myroot-sub2.slice")
+
+	// adding another sub-group to this group fails
+	_, err = rootGrp.NewSubGroup("sub3", 1)
```
Nice! This makes me wonder slightly if in the future we may limit the depth of the nesting but for now I think this is fine

mvo5 reviewed (dismissed):
Thank you

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+	// limit, oom-killer is invoked which will start killing processes. The
+	// specific behavior of which processes are killed is subject to the
+	// ExhaustionBehavior.
+	MemoryLimit quantity.Size `json:"memory-limit,omitempty"`
```
systemd modern name for this is MemoryMax. I wonder given though that we have separate behavior, whether we should use a name that is not used by systemd? otoh we it depends also what name/combination we will use in the API/command line.

anonymouse64 replied:
I have no strong opinions on the name we use here right now

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+// Group is a quota group of snaps, services or sub-groups that are all subject
+// to specific resource quotas. Current quota resource types supported are just
+// memory, but this can be expanded in the future.
+type Group struct {
```
it's a bit hard to judge fully this struct without some idea how we plan to store these into the state. Having name reference instead of full tree might be serialization friendlier but not sure.

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/gadget/quantity"
```
seems that package location is not great anymore, anyway not something to take care of immediately.

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+ *
+ */
+
+package quota
```
I suspect we might want to make this a subpackage of snap.

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+func (grp *Group) SliceFileName() string {
+	if grp.parentGroup == nil {
+		// root group name, then the slice unit is just "<name>.slice"
+		return fmt.Sprintf("%s.slice", grp.Name)
```
this needs a "snap." prefix

anonymouse64 replied:
dpme

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+	for _, parentGrpName := range grpNames {
+		fmt.Fprintf(buf, "%s-", parentGrpName)
+	}
+	fmt.Fprintf(buf, "%s.slice", grp.Name)
```
same

anonymouse64 replied:
done

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+// memory, but this can be expanded in the future.
+type Group struct {
+	// Name is the name of the quota group. Certain names are reserved and have
+	// special meaning such as user.* or system.* or snapd.*, but are otherwise
```
we need to introduce a naming.ValidateQuotaGroup to define exactly what names we allow

anonymouse64 replied:
Added

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,180 @@
+
+	buf := &bytes.Buffer{}
+	for _, parentGrpName := range grpNames {
+		fmt.Fprintf(buf, "%s-", parentGrpName)
```
for consistency with our other names we might allow '-' in quota group names which means this function needs to do appropriate quoting like we do for mounts?

anonymouse64 replied:
yes I was hoping to avoid doing the escaping of "-" for simplicity's sake, but it's certainly doable to allow "-" and escape it

anonymouse64 replied:
this is done now

pedronis reviewed (commented):
some more comments/questions about names

anonymouse64 commented:
I updated this with the new structure here and added some methods for getting the full necessary set of groups for a single group that we need in EnsureSnapServices to not be annoying and require every caller of EnsureSnapServices to pass all quota groups around,

However, I need to still address some of @pedronis' feedback here on the previous iteration, I will do that on Monday, but maybe this is enough for folks to agree on the direction of how the quotas ought to be stored in the state and manipulated

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+}
+
+// subTree recursively returns all of the sub-groups of the group
+func (grp *Group) subTree() []*Group {
```
this is probably fine because we don't expect trees to be very deep?

anonymouse64 replied:
yes that is the expectation

pedronis commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+	return subTreeList
+}
+
+func deleteGrpFromList(grps []*Group, grpToDelete *Group) []*Group {
```
givent that is is functional I suppose this should be called exlude... maybe ? (I know it would likely be called delete in a language where things are all functional)

anonymouse64 replied:
this helper actually goes away with your proposed implementation from the other PR

pedronis reviewed (commented):
thanks, direction apart  old/possible details seems reasonable, some further comments

stolowski commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+	// MemoryExhaustionBehavior is the behavior that is implemented to decide
+	// what processes to kill in the quota group when the memory limits are
+	// exhausted.
+	MemoryExhaustionBehavior MemoryExhaustionBehavior `json:"memory-exhaustion-behavior,omitempty"`
```
Maybe just OOMBehavior / oom-behavior?

anonymouse64 replied:
@pedronis thoughts on this one?

pedronis replied:
it's a bit unclear whether we should introduce this at all given it has only one possible value atm,  it's JSON is it would be easy to introduce it later when we have more than one value, and use no value as what the initial behavior was.

stolowski commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+func (grp *Group) validate() error {
+	// check that the name is a simple alphanumeric name
+	if !validGroupNameRegexp.MatchString(grp.Name) {
+		return fmt.Errorf("group name %q contains invalid characters (valid names are alphanumeric starting with a letter)", grp.Name)
```
Nitpick, maybe a separate error case for "group name must not be empty"?

stolowski commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+	return buf.String()
+}
+
+// allow only alphanumerics at least one character long, and not starting with
```
For snap names we require >=2 characters, should we require the same here?

stolowski commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+	if grp.parentGroup != nil {
+		totalAvailable := quantity.Size(0)
+		for _, child := range grp.parentGroup.subGroups {
+			totalAvailable += child.MemoryLimit
```
Nitpick, I think totalAvailable is slightly confusing, maybe `totalChildMemory`? IMHO this would read better in the line below:
`if totalChildMemory +grp.MemoryLimit > grp.parentGroup.MemoryLimit {...`

anonymouse64 replied:
I ended up going with Maciej's suggestion from the other PR, hopefully this makes sense

stolowski commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+	// parent, this is fine in systemd world, but in snapd we want unique quota
+	// groups
+	if name == grp.Name {
+		return nil, fmt.Errorf("cannot use same name %q for sub group as parent group", name)
```
That's ok but I guess we need more checks - where are we going to ensure the name is globally unique for all the groups?

anonymouse64 replied:
The intention is that there will be checks inside servicestate where we store quotas inside state.json as a map of the name to the group object, such that it will be immediately obvious whether a new user-provided name already exists as another group or not

stolowski commented on quota/quota.go:
```diff
@@ -0,0 +1,315 @@
+		return fmt.Errorf("group memory limit must be non-zero")
+	}
+
+	// TODO: probably there is a minimum amount of bytes here that is
```
Good point, I think we should investigate this in a followup.

pedronis commented on quota/quota_test.go:
```diff
@@ -0,0 +1,543 @@
+	c.Assert(qs.AllQuotaGroups(), DeepEquals, []*quota.Group{grp1, grp2})
+
+	// make a sub-group and re-add the root group - it will automatically add
+	// the sub-group without us needing to explicitly add the sub-group
```
a bit unclear to me what usage pattern would use this in particular. In the end you need to add the group in some way? my preference is just to assume that we don't support updating subgroups while adding to QuotaGroupSet

anonymouse64 replied:
This is just a unit test we likely won't use it that way. You can see how we use this function in the bigger PR but essentially it's just that we might be modifying one group in the tree, and thus we want to ensure that all groups from the tree are ensured to be updated.

I agree that I don't know in what situation changing the root group would have an effect on changing the sub-groups but I don't see how it breaks anything to do this and it's simpler code

pedronis reviewed (approved):
+1, I did my review and pushed already the changes, there was an actual bug as far as I can tell:

see  81f4074

anonymouse64 commented on quota/quota.go:
```diff
@@ -234,12 +234,6 @@ func ResolveCrossReferences(grps map[string]*Group) error {
 			}
 			grp.parentGroup = parent
 
-			// also add an internal link from the parent to this group, this
```
thanks for spotting that this was unnecessary, this was part of a previous implementation about doing the cross-checks that parents specify children and vice-versa

anonymouse64 commented on quota/quota.go:
```diff
@@ -313,6 +314,11 @@ func (s *QuotaGroupSet) AddAllNecessaryGroups(grp *Group) {
 		nextParentGrp = nextParentGrp.parentGroup
 	}
 
+	if s.grps[prevParentGrp] {
```
is there any real reason to implement this shortcut? we don't expect the trees to be large so we are not really saving any time here with this shortcut

pedronis replied:
that's true to an extent but is still enlarges a factor that then gets multiplied by all the snap services on the system potentially. But my 2nd reason for doing this is to encode the fact that we don't support mixing adding to QGS and manipulating groups at the same time,  and the reason I decided to document that is not that it let's us to the shortcut, but that once I started thinking about it , it's hard to reason exactly what you can do and can't do, so don't do that is easier to reason about.

anonymouse64 replied:
Okay, I decided to just leave this here, it doesn't really hurt anything to have it I suppose, and we should be well testing the code enough to catch any bugs anyways

anonymouse64 commented on quota/quota.go:
```diff
@@ -297,7 +297,8 @@ type QuotaGroupSet struct {
 // fully realized on a system, since all sub-branches of the full tree must
 // exist since this group may share some quota resources with the other
-// branches.
+// branches. There is no support fo manipulating group trees while
```
```suggestion
// branches. There is no support for manipulating group trees while
```

anonymouse64 reviewed (commented):
@pedronis changes look good to me, thanks for fixing that one bug, otherwise there's just a couple of clarifying questions about your changes/proposals

pedronis commented on snap/quota/quota.go:
```diff
@@ -0,0 +1,347 @@
+	// MemoryExhaustionBehavior is the behavior that is implemented to decide
+	// what processes to kill in the quota group when the memory limits are
+	// exhausted.
+	MemoryExhaustionBehavior MemoryExhaustionBehavior `json:"memory-exhaustion-behavior,omitempty"`
```
yes, let's drop this

anonymouse64 replied:
done

mvo5 commented on snap/quota/quota_test.go:
```diff
@@ -0,0 +1,601 @@
+
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/snap/quota"
+	. "gopkg.in/check.v1"
```
(nitpick^2) we usually have this above the "snapcore/snapd" imports

anonymouse64 replied:
I'll address this in the followup

mvo5 reviewed (approved):
Thank you

stolowski commented on snap/naming/validate.go:
```diff
@@ -185,3 +185,32 @@ func ValidateSecurityTag(tag string) error {
+	}
+
+	if grp[0] == '-' || grp[len(grp)-1] == '-' || strings.Contains(grp, "--") {
+		return fmt.Errorf("invalid quota group name: has invalid \"-\" sequences in it in it")
```
```suggestion
		return fmt.Errorf("invalid quota group name: has invalid \"-\" sequences in it")
```

anonymouse64 replied:
done in https://github.com/snapcore/snapd/pull/10198

stolowski commented on snap/quota/quota.go:
```diff
@@ -0,0 +1,327 @@
+
+	// save the details of this new sub-group in the parent group
+	grp.subGroups = append(grp.subGroups, subGrp)
+	grp.SubGroups = append(grp.SubGroups, name)
```
Nitpick, I found subGroups and SubGroups slightly confusing when I reached this line, maybe one should be renamed to SubGroupNames, or the non-exported one to just `groups` (since it's a member under Group struct, it's in a way obvious it is about sub-groups. But not a blocker.

stolowski commented on snap/quota/quota.go:
```diff
@@ -0,0 +1,327 @@
+	// together with their respective parent groups from the set
+	for name, grp := range grps {
+		if name != grp.Name {
+			return fmt.Errorf("group has name %q, but is referenced as %q", grp.Name, name)
```
Nitpick, this could probably be made an 'internal error:'.

stolowski commented on snap/quota/quota.go:
```diff
@@ -0,0 +1,327 @@
+			found := false
+			for _, parentChildName := range parent.SubGroups {
+				if parentChildName == grp.Name {
+					found = true
```
break here?

anonymouse64 replied:
done in https://github.com/snapcore/snapd/pull/10198

stolowski commented on snap/quota/quota.go:
```diff
@@ -0,0 +1,327 @@
+// itself.
+func (grp *Group) tree() []*Group {
+	// TODO: should we be paranoid about circular references and
+	// recursion here?
```
I think should do something about it (may come later).

anonymouse64 replied:
I added something in https://github.com/snapcore/snapd/pull/10198

stolowski reviewed (approved):
Looks good, thank you. Two small points (about a typo in error message and break statement) that would be great to address before landing, and a couple of nitpicks/remarks. Thanks!
